RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+```
+
+Method can be used to start a new Multi Test Run Finalization process, which is reprocessing all data collector attachments passed as first argument using all available processors.
+
+
+3. Introduce a new `IMultiTestRunFinalizationEventsHandler` interface:
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.Client
+{
+    /// <summary>
+    /// Interface contract for handling multi test run finalization complete events
+    /// </summary>
+    public interface IMultiTestRunFinalizationEventsHandler : ITestMessageEventHandler
+    {
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationComplete event to listeners.
+        /// </summary>
+        /// <param name="finalizationCompleteEventArgs">Finalization Complete event args.</param>
+        /// <param name="attachments">Last set of processed attachment sets.</param>
+        void HandleMultiTestRunFinalizationComplete(MultiTestRunFinalizationCompleteEventArgs finalizationCompleteEventArgs, IEnumerable<AttachmentSet> lastChunk);
+
+        /// <summary>
+        /// Dispatch FinalisedAttachments event to listeners.
+        /// </summary>
+        /// <param name="attachments">Finalised attachment sets.</param>
+        void HandleFinalisedAttachments(IEnumerable<AttachmentSet> attachments);
+
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationProgress event to listeners.
+        /// </summary>
+        /// <param name="finalizationProgressEventArgs">Finalization Progress event args.</param>
+        void HandleMultiTestRunFinalizationProgress(MultiTestRunFinalizationProgressEventArgs finalizationProgressEventArgs);
+    }
+}
+```
+Interface provides callbacks from Multi Test Run Finalization process. For every such process `HandleMultiTestRunFinalizationComplete` will be called once and will provide last chunk or all data collector attachments. During finalization process `HandleFinalisedAttachments` can be invoked several times providing data collector attachments that are already processed. Method `HandleMultiTestRunFinalizationProgress` will be invoked every time when `progressReporter` is used by `IDataCollectorAttachmentProcessor` implementation and will provide information about current collector: progress, uris and index of collector. Additionally event will contain also number of processors.
+
+
+4. Use above logic to reprocess data collector attachments for parallel test executions and VS scenarios (e.g. `Run All Tests`, `Analyze Code Coverage for All Tests`). In case of `Analyze Code Coverage for All Tests` VS will use `vstest.console` in a variation of design mode and merge all code coverage reports. VS will show full code coverage report for all test projects.
+
+5. When [dotnet test](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test) command is used to execute unit tests in a given solution, new console app `Orchestrator` will be executed.
```

Comment:
will running the Orchestrator impact the performance of a run? or do we expect the same number of processes to be spawned?

Reply:
currently all projects in solution ale executed in parallel which gives you n processes. Our change will add 2 additional processes: orchestrator and vstest.console for merging. All together we will have at least n+2 processes.

Every project in solution is starting additionally datacollector, testhost, and in case of CC also codecoverage.exe.
So in case of code coverage dotnet test we can have at least 4n + 2 processes. Maybe i still miss something

Reply:
As this part was removed, resolving it